File: internal/secrets/subkey/subkey.go
// Package subkey implements the cert-signed node HPKE sub-key and the
// client-side verification layer that turns "a trusted node HPKE pubkey" (which
// internal/secrets/seal assumes as already-trusted input) into something
// actually verified against pinned roots.
//
// It composes the already-merged building blocks rather than reinventing them:
//   - internal/secrets/seal — the HPKE envelope + ReSealToNode/OpenFromOwner.
//   - internal/pki          — the name-constrained CA chain + cert helpers.
//   - internal/clientverify — the client-side infra/SAN verification primitive.
//
// Design: docs/superpowers/specs/2026-06-10-owner-sealed-secrets-design.md §1
// (node sub-key + revocation) and §3 (delivery flow + verification chain).
//
// The pattern is RFC 9345 delegated credentials / Signal signed-prekeys: a node
// generates an X25519 HPKE sub-keypair and publishes the pubkey in a small
// SignedSubKey structure signed by its cert key (ECDSA P-256, the key type the
// node leaf cert uses — see internal/pki) with an expiry. A sealing client
// verifies that signature chains to the pinned Root CA via the node leaf cert,
// so a compromised CP can relay keys but cannot mint trust.
package subkey

import (
	"crypto/ecdsa"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"fmt"
	"time"

	"spawnery/internal/secrets/seal"
)

// DefaultValidity is the sub-key lifetime (spec §1: "Validity 72 h, rotate at
// half-life").
const DefaultValidity = 72 * time.Hour

// MaxRetained is the number of unexpired sub-key private halves a node keeps so
// a rotation mid-delivery does not fail opaquely (spec §1, roast m2).
const MaxRetained = 2

// sigDomain domain-separates the sub-key signature so a node cert signature over
// some other structure can never be replayed as a sub-key signature.
const sigDomain = "spawnery/secrets/subkey/v1"

// Errors returned by this package.
var (
	ErrBadHPKEPub   = errors.New("subkey: malformed HPKE pubkey (want 32 bytes)")
	ErrBadSig       = errors.New("subkey: signature does not verify against the node cert key")
	ErrNotYetValid  = errors.New("subkey: not yet valid (notBefore > now)")
	ErrExpired      = errors.New("subkey: expired (notAfter <= now)")
	ErrNodeMismatch = errors.New("subkey: sub-key nodeID does not match the verified cert identity")
	ErrNoSubKey     = errors.New("subkey: node has no unexpired sub-key")
	ErrNoMatch      = errors.New("subkey: no retained sub-key opens this delivery")
)

const x25519KeySize = 32

// SignedSubKey is a node's published HPKE sub-key: the raw X25519 pubkey, the
// validity window, the owning nodeID, and an ECDSA-P256 signature by the node's
// cert private key over all of those. It is relayed (untrusted) by the CP; a
// sealing client verifies Sig against the verified node leaf cert's public key.
type SignedSubKey struct {
	HPKEPub   []byte    `json:"hpke_pub"`   // raw 32-byte X25519 (seal/HPKE) pubkey
	NodeID    string    `json:"node_id"`    // must equal the cert SAN's nodeID
	NotBefore time.Time `json:"not_before"` // validity start
	NotAfter  time.Time `json:"not_after"`  // validity end (bound into delivery AAD)
	Sig       []byte    `json:"sig"`        // ECDSA-P256 ASN.1 sig over signedBytes()
}

// signedBytes is the canonical, unambiguous byte string the signature covers.
func (s SignedSubKey) signedBytes() []byte {
	return encodeFields(
		[]byte(sigDomain),
		s.HPKEPub,
		[]byte(s.NodeID),
		u64(uint64(s.NotBefore.UnixNano())),
		u64(uint64(s.NotAfter.UnixNano())),
	)
}

// KeyID is a stable short identifier for this sub-key (SHA-256 of the pubkey,
// base64url, first 16 bytes). Useful for logs and for de-duplicating retained
// keys; selection on Open is by trial-Open (see Node.OpenDelivered).
func (s SignedSubKey) KeyID() string {
	sum := sha256.Sum256(s.HPKEPub)
	return base64.RawURLEncoding.EncodeToString(sum[:16])
}

// Sign builds a SignedSubKey for hpkePub over [notBefore, notAfter) bound to
// nodeID, signed by the node's cert private key (signKey).
func Sign(signKey *ecdsa.PrivateKey, nodeID string, hpkePub []byte, notBefore, notAfter time.Time) (SignedSubKey, error) {
	if len(hpkePub) != x25519KeySize {
		return SignedSubKey{}, ErrBadHPKEPub
	}
	if signKey == nil {
		return SignedSubKey{}, errors.New("subkey: nil signing key")
	}
	s := SignedSubKey{
		HPKEPub:   append([]byte(nil), hpkePub...),
		NodeID:    nodeID,
		NotBefore: notBefore,
		NotAfter:  notAfter,
	}
	digest := sha256.Sum256(s.signedBytes())
	sig, err := ecdsa.SignASN1(rand.Reader, signKey, digest[:])
	if err != nil {
		return SignedSubKey{}, fmt.Errorf("subkey: sign: %w", err)
	}
	s.Sig = sig
	return s, nil
}

// Verify checks the sub-key's signature against certPub (the node leaf cert's
// public key). It does NOT check expiry — that is a separate, clock-injected
// step (see VerifyNodeForSealing / Valid).
func (s SignedSubKey) Verify(certPub *ecdsa.PublicKey) error {
	if certPub == nil {
		return errors.New("subkey: nil cert public key")
	}
	if len(s.HPKEPub) != x25519KeySize {
		return ErrBadHPKEPub
	}
	digest := sha256.Sum256(s.signedBytes())
	if !ecdsa.VerifyASN1(certPub, digest[:], s.Sig) {
		return ErrBadSig
	}
	return nil
}

// Valid reports whether now is within [NotBefore, NotAfter). It is the expiry
// half of verification, kept separate from the signature check so callers can
// inject now.
func (s SignedSubKey) Valid(now time.Time) error {
	if now.Before(s.NotBefore) {
		return ErrNotYetValid
	}
	if !now.Before(s.NotAfter) {
		return ErrExpired
	}
	return nil
}

// Node is the node-side holder of HPKE sub-keys: it generates and re-signs
// sub-keys with the node cert key, retains up to MaxRetained unexpired private
// halves, and opens delivered ciphertexts by trial-Open across them.
type Node struct {
	signKey  *ecdsa.PrivateKey
	nodeID   string
	validity time.Duration
	retained []retained // newest first
}

type retained struct {
	signed SignedSubKey
	priv   []byte // raw 32-byte X25519 private half
}

// NewNode constructs a holder for nodeID signing with signKey (the node cert
// private key) and the given sub-key validity. Call Rotate to mint the first
// sub-key. If validity <= 0, DefaultValidity is used.
func NewNode(signKey *ecdsa.PrivateKey, nodeID string, validity time.Duration) *Node {
	if validity <= 0 {
		validity = DefaultValidity
	}
	return &Node{signKey: signKey, nodeID: nodeID, validity: validity}
}

// Rotate generates a fresh X25519 sub-keypair, signs it for [now, now+validity),
// retains its private half, prunes expired halves, caps retention at
// MaxRetained, and returns the newly minted SignedSubKey to publish.
func (n *Node) Rotate(now time.Time) (SignedSubKey, error) {
	pub, priv, err := seal.NodeKeyPair()
	if err != nil {
		return SignedSubKey{}, err
	}
	signed, err := Sign(n.signKey, n.nodeID, pub, now, now.Add(n.validity))
	if err != nil {
		return SignedSubKey{}, err
	}
	// Newest first.
	n.retained = append([]retained{{signed: signed, priv: priv}}, n.retained...)
	n.prune(now)
	return signed, nil
}

// prune drops expired retained halves and caps the rest (newest-first) at
// MaxRetained.
func (n *Node) prune(now time.Time) {
	kept := n.retained[:0]
	for _, r := range n.retained {
		if r.signed.Valid(now) == nil && len(kept) < MaxRetained {
			kept = append(kept, r)
		}
	}
	n.retained = kept
}

// Current returns the freshest unexpired SignedSubKey to publish, if any.
func (n *Node) Current(now time.Time) (SignedSubKey, bool) {
	for _, r := range n.retained {
		if r.signed.Valid(now) == nil {
			return r.signed, true
		}
	}
	return SignedSubKey{}, false
}

// NeedsRotation reports whether the freshest sub-key is past its half-life (or
// there is no unexpired sub-key) — the cue to call Rotate (spec §1: rotate at
// half-life).
func (n *Node) NeedsRotation(now time.Time) bool {
	cur, ok := n.Current(now)
	if !ok {
		return true
	}
	halfLife := cur.NotBefore.Add(cur.NotAfter.Sub(cur.NotBefore) / 2)
	return !now.Before(halfLife)
}

// Retained returns the count of unexpired retained sub-key private halves.
func (n *Node) Retained(now time.Time) int {
	c := 0
	for _, r := range n.retained {
		if r.signed.Valid(now) == nil {
			c++
		}
	}
	return c
}

// OpenDelivered opens a CP-relayed NodeSealed by trial-Open across the retained,
// unexpired sub-key private halves: for each, it reconstructs the expected
// in-flight AAD with THAT sub-key's NotAfter (and this node's NodeID) and calls
// seal.OpenFromOwner. The first that opens wins, so a rotation mid-delivery
// (two concurrent sub-keys) does not fail opaquely (spec §1, roast m2).
//
// baseAAD carries the per-delivery context the node knows out-of-band (spawnId,
// generation, version, deliveryId); NodeID and NotAfter are supplied per-key so
// the node need not know out-of-band which sub-key the client sealed to.
//
// Caller still MUST enforce the stateful §3 checks OpenFromOwner cannot:
// version-monotonic and deliveryId-once (see seal.OpenFromOwner).
func (n *Node) OpenDelivered(sealed *seal.NodeSealed, baseAAD seal.InFlightAAD, now time.Time) ([]byte, error) {
	tried := 0
	for _, r := range n.retained {
		if r.signed.Valid(now) != nil {
			continue
		}
		tried++
		aad := baseAAD
		aad.NodeID = n.nodeID
		aad.NotAfter = r.signed.NotAfter
		pt, err := seal.OpenFromOwner(sealed, r.priv, aad, now)
		if err == nil {
			return pt, nil
		}
	}
	if tried == 0 {
		return nil, ErrNoSubKey
	}
	return nil, ErrNoMatch
}

// encodeFields produces an unambiguous length-prefixed concatenation of its
// parts (each prefixed with an 8-byte big-endian length), so distinct field
// tuples never collide into the same signed byte string. (Mirrors the encoder
// in internal/secrets/seal; kept local to avoid widening that package's API.)
func encodeFields(parts ...[]byte) []byte {
	n := 0
	for _, p := range parts {
		n += 8 + len(p)
	}
	out := make([]byte, 0, n)
	var lb [8]byte
	for _, p := range parts {
		binary.BigEndian.PutUint64(lb[:], uint64(len(p)))
		out = append(out, lb[:]...)
		out = append(out, p...)
	}
	return out
}

func u64(v uint64) []byte {
	var b [8]byte
	binary.BigEndian.PutUint64(b[:], v)
	return b[:]
}

File: web/src/keys/subkey.ts
/**
 * Node HPKE sub-key: TypeScript mirror of internal/secrets/subkey (Go).
 *
 * Implements:
 *   - SignedSubKey — the JSON structure published by a node
 *   - verifySignedSubKey — ECDSA-P256 sig check + expiry check
 *   - verifyNodeForSealing — full verification chain (cert + sub-key) returning
 *     a trusted HPKE pubkey for use with seal/reseal operations
 *
 * Design: docs/superpowers/specs/2026-06-10-owner-sealed-secrets-design.md §1 §3
 *
 * Node cert profile: P-256 leaf, SAN = <nodeId>.<accountId>.<class>.nodes.spawnery.internal
 *
 * WM10: UnixNano timestamps are int64 — must use BigInt in encodeFields.
 */

import { derToP1363 } from "./der";
import {
  verifyCertChain,
  importCertPubKey,
  parseSANIdentity,
  ParsedCert,
} from "./x509";

// ── Encoding helpers (mirror of subkey.go + seal.go encodeFields) ────────────

const enc = new TextEncoder();

function encodeFields(...parts: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const p of parts) total += 8 + p.length;
  const out = new Uint8Array(total);
  let off = 0;
  const view = new DataView(out.buffer);
  for (const p of parts) {
    view.setBigUint64(off, BigInt(p.length), false);
    off += 8;
    out.set(p, off);
    off += p.length;
  }
  return out;
}

/** Encode a BigInt as 8-byte big-endian. Must use BigInt for UnixNano (WM10). */
function u64big(v: bigint): Uint8Array {
  const b = new Uint8Array(8);
  new DataView(b.buffer).setBigUint64(0, v, false);
  return b;
}

/** Helpers to convert from base64. */
function fromBase64(b64: string): Uint8Array {
  const bin = atob(b64);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

// ── SignedSubKey ──────────────────────────────────────────────────────────────

/** Mirror of subkey.go SignedSubKey (JSON-deserialized from the CP). */
export interface SignedSubKey {
  hpke_pub:   string; // base64 raw 32-byte X25519 pubkey
  node_id:    string;
  not_before: string; // RFC 3339 / ISO 8601
  not_after:  string; // RFC 3339 / ISO 8601
  sig:        string; // base64 ASN.1 DER ECDSA-P256 signature
}

// sigDomain must match Go's subkey.go const sigDomain.
const sigDomain = "spawnery/secrets/subkey/v1";

/**
 * Compute the canonical byte string that the sub-key signature covers.
 * Mirror of SignedSubKey.signedBytes() in Go.
 *
 * encodeFields(sigDomain, HPKEPub, NodeID, u64(NotBefore.UnixNano), u64(NotAfter.UnixNano))
 */
function signedBytes(sk: SignedSubKey): Uint8Array {
  const hpkePub   = fromBase64(sk.hpke_pub);
  const notBefore = BigInt(new Date(sk.not_before).getTime()) * 1_000_000n; // ms → ns
  const notAfter  = BigInt(new Date(sk.not_after).getTime()) * 1_000_000n;
  return encodeFields(
    enc.encode(sigDomain),
    hpkePub,
    enc.encode(sk.node_id),
    u64big(notBefore),
    u64big(notAfter),
  );
}

/**
 * Verify a SignedSubKey's ECDSA-P256 signature against certPub (the leaf cert
 * public key), and check that now is within [not_before, not_after).
 * Throws on any failure.
 */
export async function verifySignedSubKey(
  sk: SignedSubKey,
  certPub: CryptoKey,
  now: Date,
): Promise<void> {
  // Expiry check.
  const notBefore = new Date(sk.not_before);
  const notAfter  = new Date(sk.not_after);
  if (now < notBefore) throw new Error(`subkey: not yet valid (not_before=${sk.not_before})`);
  if (now >= notAfter)  throw new Error(`subkey: expired (not_after=${sk.not_after})`);

  // Signature check.
  const sigDER = fromBase64(sk.sig);
  const sigP1363 = derToP1363(sigDER);
  const msg = signedBytes(sk);
  const ok = await crypto.subtle.verify(
    { name: "ECDSA", hash: "SHA-256" },
    certPub,
    sigP1363.buffer.slice(sigP1363.byteOffset, sigP1363.byteOffset + sigP1363.byteLength) as ArrayBuffer,
    msg.buffer.slice(msg.byteOffset, msg.byteOffset + msg.byteLength) as ArrayBuffer,
  );
  if (!ok) throw new Error("subkey: signature does not verify against the node cert key");
}

// ── VerifyNodeForSealing ──────────────────────────────────────────────────────

/**
 * Verified node identity returned by verifyNodeForSealing.
 * Mirrors pki.Identity in Go.
 */
export interface NodeIdentity {
  nodeId:    string;
  accountId: string;
  nodeClass: string; // "cloud" | "self-hosted"
}

/**
 * Client-side sealing expectation (mirrors subkey.Expectation / clientverify.Expectation in Go).
 * tenancy = "cloud" | "self-hosted". For self-hosted, accountId is checked.
 */
export interface SealingExpectation {
  tenancy:   "cloud" | "self-hosted";
  accountId?: string;
}

/**
 * Full verification chain before sealing (spec §3 step 2). In order:
 *
 *   1. node cert chains to pinned rootPEM + SAN matches expect
 *   2. sub-key nodeID matches verified cert identity
 *   3. sub-key signature chains to cert key
 *   4. sub-key is unexpired
 *
 * Returns the trusted HPKE X25519 pubkey (raw 32 bytes) and the verified node
 * identity, or throws on any failure.
 *
 * certChainPEM is the CP-relayed leaf+chain PEM; rootPEM is the client-pinned
 * Root CA. If certChainPEM is empty (dev/insecure mode), step 1 is skipped and
 * the sub-key pubkey is returned trusted with a synthetic "dev" identity.
 *
 * subkeyJSON is the raw JSON string from GetSpawnNodeKeyResponse.signed_subkey.
 * now is the current time (injectable for testing).
 */
export async function verifyNodeForSealing(
  certChainPEM: string,
  rootPEM: string,
  subkeyJSON: string,
  expect: SealingExpectation,
  now: Date,
): Promise<{ hpkePub: Uint8Array; identity: NodeIdentity }> {
  const sk: SignedSubKey = JSON.parse(subkeyJSON);

  // Dev/insecure mode: no cert chain → skip chain verification.
  if (!certChainPEM) {
    return {
      hpkePub:  fromBase64(sk.hpke_pub),
      identity: { nodeId: sk.node_id, accountId: "dev", nodeClass: "cloud" },
    };
  }

  // 1. Verify cert chain against pinned root + extract identity from SAN.
  const leaf = await verifyCertChain(certChainPEM, rootPEM);
  const identity = parseSANIdentity(leaf.sanDNS);

  // Tenancy check.
  if (identity.nodeClass !== expect.tenancy) {
    throw new Error(`subkey: expected tenancy ${JSON.stringify(expect.tenancy)}, got ${JSON.stringify(identity.nodeClass)}`);
  }
  if (expect.tenancy === "self-hosted") {
    if (!expect.accountId) throw new Error("subkey: accountId required for self-hosted expectation");
    if (identity.accountId !== expect.accountId) {
      throw new Error(`subkey: self-hosted node bound to ${JSON.stringify(identity.accountId)}, not ${JSON.stringify(expect.accountId)}`);
    }
  }

  // 2. Sub-key nodeID must match verified cert identity.
  if (sk.node_id !== identity.nodeId) {
    throw new Error(`subkey: sub-key node_id=${JSON.stringify(sk.node_id)} != cert nodeId=${JSON.stringify(identity.nodeId)}`);
  }

  // 3 + 4. Verify sub-key signature + expiry.
  const certPub = await importCertPubKey(leaf);
  await verifySignedSubKey(sk, certPub, now);

  return { hpkePub: fromBase64(sk.hpke_pub), identity };
}

// Export ParsedCert for callers that want the leaf cert separately.
export type { ParsedCert };
